third_party/move/move-vm/types/src/utils/union_find.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#[derive(Debug)]
+pub struct UnionFind {
+    parent: Vec<u32>,
+    rank: Vec<u8>,
+    num_sets: usize,
+    capacity: usize,
+}
+
+impl UnionFind {
+    /// Creates a new UnionFind instance with `num_elements` and `capacity`
+    pub fn new_with_size_and_capacity(
+        num_elements: usize,
+        capacity: usize,
+    ) -> Result<Self, &'static str> {
```

Comment:
Why error is a string? Can we use `anyhow::Result`? Also seems like invariants, isn't it better to panic in this case and make these asserts?